### PR TITLE
[Perf] Thread local storage for range-for reductions on GPUs

### PR DESCRIPTION
**Pull request by yuanming-hu**
Related issue = #576

 - Support GPU range-for TLS reduction (1.5x faster compared w/o TLS, and achieves 95% theoretical performance)
 - Make TLS buffer allocation adaptive

[[Click here for the format server]](http://kun.csail.mit.edu:31415/)

----

### CODE REVIEW

File: taichi/runtime/llvm/runtime.cpp [hunk 1]
```diff
@@ -1046,7 +1047,7 @@ struct range_task_helper_context {
 
 void parallel_range_for_task(void *range_context, int task_id) {
   auto ctx = *(range_task_helper_context *)range_context;
-  char tls_buffer[taichi_tls_buffer_size];
+  alignas(8) char tls_buffer[ctx.tls_size];
```
Comment (k-ye) [inline taichi/runtime/llvm/runtime.cpp:1050]: Wow I didn't even know we can do variable length array nowadays. Do you know if this is supported on C++17 now? (I did a quick search and found that while this is supported in C99, the support of VLA in C++ remains vague.. I tested with `clang` and it's working)

File: taichi/transforms/make_thread_local.cpp [hunk 2]
```diff
@@ -127,10 +131,12 @@ void make_thread_local_offload(OffloadedStmt *offload) {
...
     }
-    tls_offset += tls_stride;
-    if (tls_offset >= taichi_tls_buffer_size)
-      break;  // Do not overflow TLS buffer.  TODO: make it adaptive
+
+    // allocate storage for the TLS variable
+    tls_offset += dtype_size;
   }
+
+  offload->tls_size = tls_offset;
```
Comment (k-ye) [inline taichi/transforms/make_thread_local.cpp:139]: `std::max(1, tls_offset)`?

File: taichi/program/kernel.cpp [hunk 3]
```diff
@@ -75,8 +75,7 @@ void Kernel::lower(bool lower_access) {  // TODO: is a "Lowerer" class necessary
         (is_evaluator && !config.print_evaluator_ir))
       verbose = false;
 
-    // For now we only enable TLS on CPUs
-    bool make_thread_local = arch_is_cpu(arch);
+    bool make_thread_local = true;
```
Comment (k-ye) [inline taichi/program/kernel.cpp:78]: Only CPU or CUDA?

File: taichi/runtime/llvm/runtime.cpp [hunk 4]
```diff
@@ -1105,12 +1108,21 @@ void cpu_parallel_range_for(Context *context,
 void gpu_parallel_range_for(Context *context,
                             int begin,
                             int end,
-                            RangeForTaskFunc *func) {
+                            range_for_xlogue prologue,
+                            RangeForTaskFunc *func,
+                            range_for_xlogue epilogue,
+                            const std::size_t tls_size) {
   int idx = thread_idx() + block_dim() * block_idx() + begin;
+  alignas(8) char tls_buffer[tls_size];
```
Comment (k-ye) [inline taichi/runtime/llvm/runtime.cpp:1116]: See if I understand this correctly. 

* `tls_buffer` is still a thread local buffer, pretty much like adstack? (Previously I thought TLS buffer was `__shared__`.)
* There is no warp level reduction in this PR yet. The saving of global atomics purely comes from grid-strided loops? I.e. if each thread still only handles one coordinate, then there's no difference?
* In the epilogue, we will do the warp-level reduction on these TLS buffers. By that time we need `__synchthreads()` to make sure all works in `body` are done.

File: taichi/runtime/llvm/runtime.cpp (hunk 1 above)
Comment (yuanming-hu) [reply taichi/runtime/llvm/runtime.cpp:1050]: I'm not sure if that's in C++17 TBH. However, LLVM `alloca` does support allocating arrays with the number of elements being a variable (instruction) :-)

File: taichi/program/kernel.cpp (hunk 3 above)
Comment (yuanming-hu) [reply taichi/program/kernel.cpp:78]: Yeah now we support both CPU and CUDA.

File: taichi/runtime/llvm/runtime.cpp (hunk 4 above)
Comment (yuanming-hu) [reply taichi/runtime/llvm/runtime.cpp:1116]: > * `tls_buffer` is still a thread local buffer, pretty much like adstack? (Previously I thought TLS buffer was `__shared__`.)

“Thread local” storage and shared memory (“block local”) storage are actually two different types of storage. On CUDA and Metal we need to support both.
 1. TLS is allocated as thread registers
 2. Block local storage (shared memory) is allocated per block

I should change the name to `Local Storage`.




> * There is no warp level reduction in this PR yet. The saving of global atomics purely comes from grid-strided loops? I.e. if each thread still only handles one coordinate, then there's no difference?

Right, I find that we can achieve 95% peak without warp-level reductions in this test case. You are right, we should try to add wrap-level reduction in the future - that would be helpful for smaller arrays.



> * In the epilogue, we will do the warp-level reduction on these TLS buffers. By that time we need `__synchthreads()` to make sure all works in `body` are done.

Actually, if that's just TLS + warp-level reduction, then `__synthreads()` is not needed. If you do use shared memory (which is not in this PR), then we need `__synthreads()`.

File: taichi/program/kernel.cpp (hunk 3 above)
Comment (k-ye) [reply taichi/program/kernel.cpp]: Ah, i meant to ask if this could break on other platforms like GL and Metal..

Comment (yuanming-hu) [reply taichi/program/kernel.cpp]: Oh we have this line above :-)
```
if (arch_is_cpu(arch) || arch == Arch::cuda) {
```

**Pull request merged**